### PR TITLE
RPC refresh

### PR DESCRIPTION
**Pull request by n8maninger**
Add explicit refresh RPC for more efficient and correct contract refreshes

### CODE REVIEW

File: rhp/v4/rhp.go [hunk 1]
```diff
@@ -677,3 +727,30 @@ func NewRenewal(fc types.V2FileContract, prices HostPrices, rp RPCRenewContractP
...
+
+// RefreshContract creates a new contract renewal from an existing contract revision
+func RefreshContract(fc types.V2FileContract, prices HostPrices, rp RPCRefreshContractParams) (renewal types.V2FileContractRenewal) {
+	// clear the old contract
+	renewal.FinalRevision = fc
+	renewal.FinalRevision.RevisionNumber = types.MaxRevisionNumber
+	renewal.FinalRevision.Filesize = 0
+	renewal.FinalRevision.FileMerkleRoot = types.Hash256{}
+	renewal.FinalRevision.RenterSignature = types.Signature{}
+	renewal.FinalRevision.HostSignature = types.Signature{}
```
Comment (lukechampine) [inline rhp/v4/rhp.go:739]: I don't think it's necessary to clear these fields; consensus should treat it as a valid resolution regardless.

File: rhp/v4/rhp.go [hunk 2]
```diff
@@ -265,6 +266,39 @@ type (
 		TransactionSet []types.V2Transaction `json:"transactionSet"`
 	}
 
+	// RPCRefreshContractParams includes the contract details required to create
```
Comment (peterjan) [inline rhp/v4/rhp.go:269]: `"required to refresh a contract."`?

File: rhp/v4/validation.go [hunk 3]
```diff
@@ -149,6 +151,38 @@ func (req *RPCRenewContractRequest) Validate(pk types.PublicKey, tip types.Chain
 	}
 }
 
+// Validate validates a renew contract request. Prices are not validated
```
Comment (peterjan) [inline rhp/v4/validation.go:154]:
```suggestion
// Validate validates a refresh contract request. Prices are not validated
```

File: rhp/v4/validation.go [hunk 4]
```diff
@@ -149,6 +151,38 @@ func (req *RPCRenewContractRequest) Validate(pk types.PublicKey, tip types.Chain
...
+	// host's locked collateral and the contract duration
+	minRenterAllowance := MinRenterAllowance(hp, expirationHeight-req.Prices.TipHeight, req.Refresh.Collateral)
+
+	switch {
+	case req.Refresh.Allowance.IsZero():
+		return errors.New("allowance is zero")
+	case req.Refresh.Collateral.Cmp(maxCollateral) > 0:
+		return fmt.Errorf("collateral %v exceeds max collateral %v", req.Refresh.Collateral, maxCollateral)
+	case req.Refresh.Allowance.Cmp(minRenterAllowance) < 0:
+		return fmt.Errorf("allowance %v is less than minimum %v for collateral", req.Refresh.Allowance, minRenterAllowance)
```
Comment (peterjan) [inline rhp/v4/validation.go:180]: I'm guessing "for collateral" is a typo in the error message and it's meant to be "allowance x is less than min allowance y". It's repeated three times in the repo.

File: rhp/v4/validation.go [hunk 5]
```diff
@@ -149,6 +151,38 @@ func (req *RPCRenewContractRequest) Validate(pk types.PublicKey, tip types.Chain
...
+
+	// validate the contract fields
+	hp := req.Prices
+	// calculate the minimum allowance required for the contract based on the
+	// host's locked collateral and the contract duration
+	minRenterAllowance := MinRenterAllowance(hp, expirationHeight-req.Prices.TipHeight, req.Refresh.Collateral)
+
+	switch {
+	case req.Refresh.Allowance.IsZero():
+		return errors.New("allowance is zero")
```
Comment (peterjan) [inline rhp/v4/validation.go:176]: We have two styles of error messages. Sometimes we state the value and indicate it's invalid, while other times we specify that 'value X must adhere to condition Y,' and explain that the condition was not met. I prefer the latter, where we would error out here with "allowance must be greater than zero". I'm going to get shamed for this `nit` but I don't care.

**Review by lukechampine — APPROVED** [review]

**Pull request merged**